lib/saml/settings_service.rb

```diff
@@ -1,28 +1,43 @@
 # frozen_string_literal: true
 require 'memoist'
+require 'sentry_logging'
 
 module SAML
   # This class is responsible for putting together a complete ruby-saml
   # SETTINGS object, meaning, our static SP settings + the IDP settings
   # which must be fetched once and only once via IDP metadata.
   class SettingsService
     class << self
+      include SentryLogging
       extend Memoist
 
       METADATA_RETRIES = 3
       OPEN_TIMEOUT = 2
       TIMEOUT = 15
 
       def saml_settings
+        if @fetch_attempted.nil? || metadata_successfully_retrieved?
+          merged_saml_settings
+        else
+          # retrieve was attempted but failed, blow away memoized data and try again
+          merged_saml_settings(true)
```

Comment:
Also, it seems incorrect that the saml_settings method knows so much about the retry logic in the `metadata` method. I think I'd be inclined to have this method just blindly call `merged_saml_settings`, which would in turn blindly call `metadata` and all of the retry logic could be handled at that level.